### PR TITLE
Add Entropy Adaptive Fine Tuning

### LINKED ISSUE #4795: [Feature Request] Entropy-Adaptive Fine-Tuning
### Feature request

code:

https://github.com/PRIS-CV/EAFT

paper:

https://huggingface.co/papers/2601.02151

### Motivation

Looks promising!

### Your contribution

I can't commit to anything at the moment

### PR DESCRIPTION
**Pull request by electroglyph**
from the paper: https://huggingface.co/papers/2601.02151

the implementation is 100% based on (and credited to):
https://github.com/hiyouga/LlamaFactory/compare/main...ymxyll:LlamaFactory-EAFT:feature/eaft

closes #4795

use like so:

```python
trainer = SFTTrainer(
    args = SFTConfig(
        loss_type = "eaft",
        eaft_alpha = 1.0, # default
    ),
)
```

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ x] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Introduces a new loss implementation and wires it into `SFTTrainer`, changing training-time loss computation and adding a new config knob; incorrect weighting/shape handling could affect training dynamics and metrics.
> 
> **Overview**
> Adds support for *Entropy-Adaptive Fine-Tuning* by introducing an `eaft` `loss_type` in `SFTTrainer`, including a new `eaft_loss_func` that applies entropy-based per-token weighting (controlled via `SFTConfig.eaft_alpha`) and disallows user-provided `compute_loss_func` when enabled.
> 
> Extends `SFTConfig` to document the new loss type and expose `eaft_alpha`, and adds a new test suite (`tests/test_eaft.py`) covering the standalone EAFT loss (including `alpha=0` equivalence to standard CE) and a short end-to-end training run plus initialization error handling.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 97270685beeae100df378c49e8b421d311311e24. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
thanks for the pr, I think that the reviews from cursor are legit, + can you make sure to make this contribution consistent with the codebase, like the test should be in the sft trainer test file, etc
